File: scripts/solar/w3ts/handles/effect.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor
local __TS__New = ____lualib.__TS__New
local __TS__Delete = ____lualib.__TS__Delete
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["10"] = 3,["11"] = 3,["12"] = 6,["13"] = 6,["14"] = 6,["15"] = 6,["16"] = 24,["17"] = 25,["18"] = 6,["19"] = 27,["20"] = 6,["22"] = 28,["24"] = 29,["25"] = 6,["27"] = 30,["30"] = 24,["36"] = 12,["37"] = 13,["39"] = 15,["41"] = 18,["42"] = 19,["52"] = 178,["53"] = 179,["54"] = 180,["56"] = 182,["58"] = 168,["59"] = 169,["60"] = 170,["61"] = 171,["71"] = 197,["72"] = 198,["73"] = 199,["75"] = 201,["77"] = 188,["78"] = 189,["79"] = 190,["89"] = 216,["91"] = 207,["92"] = 208,["93"] = 209,["103"] = 231,["104"] = 232,["105"] = 233,["107"] = 235,["109"] = 222,["110"] = 223,["111"] = 224,["121"] = 250,["123"] = 241,["124"] = 242,["125"] = 243,["135"] = 265,["137"] = 256,["138"] = 257,["139"] = 258,["149"] = 281,["151"] = 271,["152"] = 272,["153"] = 273,["154"] = 274,["164"] = 296,["166"] = 287,["167"] = 288,["168"] = 289,["178"] = 311,["180"] = 302,["181"] = 303,["182"] = 304,["187"] = 154,["188"] = 155,["189"] = 154,["190"] = 161,["191"] = 162,["192"] = 161,["193"] = 317,["194"] = 318,["195"] = 319,["196"] = 320,["197"] = 317,["198"] = 326,["199"] = 327,["200"] = 328,["201"] = 329,["203"] = 331,["204"] = 332,["206"] = 334,["207"] = 326,["208"] = 340,["209"] = 341,["210"] = 342,["211"] = 340,["212"] = 348,["213"] = 349,["214"] = 348,["215"] = 355,["216"] = 356,["217"] = 357,["218"] = 355,["219"] = 363,["220"] = 364,["221"] = 363,["222"] = 370,["223"] = 371,["224"] = 372,["225"] = 373,["226"] = 374,["227"] = 375,["228"] = 370,["229"] = 381,["230"] = 382,["231"] = 383,["232"] = 381,["233"] = 386,["234"] = 387,["235"] = 386});
local ____exports = {}
local ____handle = require("solar.w3ts.handles.handle")
local Handle = ____handle.Handle
____exports.Effect = __TS__Class()
local Effect = ____exports.Effect
Effect.name = "Effect"
__TS__ClassExtends(Effect, Handle)
function Effect.prototype.____constructor(self, modelName, a, b)
    if Handle:initFromHandle() then
        Handle.prototype.____constructor(self)
    elseif type(a) == "number" and type(b) == "number" then
        Handle.prototype.____constructor(
            self,
            AddSpecialEffect(modelName, a, b)
        )
    elseif type(a) ~= "number" and type(b) == "string" then
        Handle.prototype.____constructor(
            self,
            AddSpecialEffectTarget(modelName, a.handle, b)
        )
    end
end
__TS__SetDescriptor(
    Effect.prototype,
    "data",
    {
        get = function(self)
            if not self._data then
                self._data = {}
            end
            return self._data
        end,
        set = function(self, value)
            self._data = value
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectSize",
    {
        get = function(self)
            local size = self._data.size
            if not size then
                size = EXGetEffectSize(self.handle)
            end
            return size
        end,
        set = function(self, size)
            EXSetEffectSize(self.handle, size)
            SetPariticle2Size(self.handle, size)
            self._data.size = size
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectZ",
    {
        get = function(self)
            local z = self._data.z
            if not z then
                z = EXGetEffectZ(self.handle)
            end
            return z
        end,
        set = function(self, Z)
            EXSetEffectZ(self.handle, Z)
            self._data.z = Z
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectSpeed",
    {
        get = function(self)
            return self._data.Speed
        end,
        set = function(self, Speed)
            EXSetEffectSpeed(self.handle, Speed)
            self._data.Speed = Speed
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectColor",
    {
        get = function(self)
            local Color = self._data.color
            if not Color then
                Color = EXGetEffectColor(self.handle)
            end
            return Color
        end,
        set = function(self, Color)
            EXSetEffectColor(self.handle, Color)
            self._data.color = Color
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectFogVisible",
    {
        get = function(self)
            return self._data.fogvisible
        end,
        set = function(self, fogvisible)
            EXSetEffectFogVisible(self.handle, fogvisible)
            self._data.fogvisible = fogvisible
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectVisible",
    {
        get = function(self)
            return self._data.visible
        end,
        set = function(self, visible)
            EXSetEffectVisible(self.handle, visible)
            self._data.visible = visible
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectRotateZ",
    {
        get = function(self)
            return self._data.rotateZ
        end,
        set = function(self, z)
            self:reset()
            EXEffectMatRotateZ(self.handle, z)
            self._data.rotateZ = z
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectRotateX",
    {
        get = function(self)
            return self._data.rotateX
        end,
        set = function(self, x)
            EXEffectMatRotateX(self.handle, x)
            self._data.rotateX = x
        end
    },
    true
)
__TS__SetDescriptor(
    Effect.prototype,
    "effectRotateY",
    {
        get = function(self)
            return self._data.rotateY
        end,
        set = function(self, y)
            EXEffectMatRotateY(self.handle, y)
            self._data.rotateY = y
        end
    },
    true
)
function Effect.createEffect(self, modelName, x, y)
    return __TS__New(____exports.Effect, modelName, x, y)
end
function Effect.createEffectTarget(self, modelName, a, b)
    return __TS__New(____exports.Effect, modelName, a, b)
end
function Effect.prototype.setEffectXY(self, x, y)
    EXSetEffectXY(self.handle, x, y)
    self._data.x = x
    self._data.y = y
end
function Effect.prototype.getEffectXY(self)
    local xy = {x = self._data.x, y = self._data.y}
    if not xy.x then
        xy.x = EXGetEffectX(self.handle)
    end
    if not xy.y then
        xy.y = EXGetEffectY(self.handle)
    end
    return xy
end
function Effect.prototype.setBindEffect(self, han, node)
    BindEffect(han, node, self.handle)
    self._data.binding = han
end
function Effect.prototype.getBindEffect(self)
    return self._data.binding
end
function Effect.prototype.unBindEffect(self)
    UnBindEffect(self.handle)
    self._data.binding = nil
end
function Effect.prototype.setEffectAnimation(self, aon)
    EXSetEffectAnimation(self.handle, aon)
end
function Effect.prototype.reset(self)
    EXEffectMatReset(self.handle)
    self._data.size = 1
    self._data.rotateX = 0
    self._data.rotateY = 0
    self._data.rotateZ = 0
end
function Effect.prototype.destroy(self)
    __TS__Delete(self, "_data")
    DestroyEffect(self.handle)
end
function Effect.fromHandle(self, handle)
    return self:getObject(handle)
end
return ____exports

File: src/solar/w3ts/handles/effect.ts
/** @noSelfInFile **/

import {Handle} from "./handle";
import {Widget} from "./widget";

export class Effect extends Handle<effect> {

    /**特效对象 */
    private _data: any

    get data(): any {
        if (!this._data) {
            this._data = {}
        }
        return this._data;
    }

    set data(value: any) {
        this._data = value;
    }

    constructor(modelName: string, x: number, y: number);
    constructor(modelName: string, targetWidget: Widget, attachPointName: string);
    constructor(modelName: string, a: number | Widget, b: number | string) {
        if (Handle.initFromHandle()) {
            super();
        } else if (typeof a === "number" && typeof b === "number") {
            super(AddSpecialEffect(modelName, a, b));
        } else if (typeof a !== "number" && typeof b === "string") {
            super(AddSpecialEffectTarget(modelName, a.handle, b));
        }
    }

    // public get scale() {
    //   return BlzGetSpecialEffectScale(this.handle);
    // }
    //
    // public set scale(scale: number) {
    //   BlzSetSpecialEffectScale(this.handle, scale);
    // }

    /**
     * Warning: asynchronous
     */
    // public get x() {
    //   return BlzGetLocalSpecialEffectX(this.handle);
    // }
    //
    // public set x(x: number) {
    //   BlzSetSpecialEffectX(this.handle, x);
    // }

    /**
     * Warning: asynchronous
     */
    // public get y() {
    //   return BlzGetLocalSpecialEffectY(this.handle);
    // }
    //
    // public set y(y: number) {
    //   BlzSetSpecialEffectY(this.handle, y);
    // }
    //
    // /**
    //  * Warning: asynchronous
    //  */
    // public get z() {
    //   return BlzGetLocalSpecialEffectZ(this.handle);
    // }
    //
    // public set z(z: number) {
    //   BlzSetSpecialEffectZ(this.handle, z);
    // }
    //
    // public addSubAnimation(subAnim: subanimtype) {
    //   BlzSpecialEffectAddSubAnimation(this.handle, subAnim);
    // }
    //
    // public clearSubAnimations() {
    //   BlzSpecialEffectClearSubAnimations(this.handle);
    // }

    // public playAnimation(animType: animtype) {
    //   BlzPlaySpecialEffect(this.handle, animType);
    // }
    //
    // public playWithTimeScale(animType: animtype, timeScale: number) {
    //   BlzPlaySpecialEffectWithTimeScale(this.handle, animType, timeScale);
    // }
    //
    // public removeSubAnimation(subAnim: subanimtype) {
    //   BlzSpecialEffectRemoveSubAnimation(this.handle, subAnim);
    // }
    //
    // public resetScaleMatrix() {
    //   BlzResetSpecialEffectMatrix(this.handle);
    // }
    //
    // public setAlpha(alpha: number) {
    //   BlzSetSpecialEffectAlpha(this.handle, alpha);
    // }
    //
    // public setColor(red: number, green: number, blue: number) {
    //   BlzSetSpecialEffectColor(this.handle, red, green, blue);
    // }
    //
    // public setColorByPlayer(whichPlayer: MapPlayer) {
    //   BlzSetSpecialEffectColorByPlayer(this.handle, whichPlayer.handle);
    // }
    //
    // public setHeight(height: number) {
    //   BlzSetSpecialEffectHeight(this.handle, height);
    // }
    //
    // public setOrientation(yaw: number, pitch: number, roll: number) {
    //   BlzSetSpecialEffectOrientation(this.handle, yaw, pitch, roll);
    // }

    // public setPitch(pitch: number) {
    //   BlzSetSpecialEffectPitch(this.handle, pitch);
    // }
    //
    // public setPoint(p: Point) {
    //   BlzSetSpecialEffectPositionLoc(this.handle, p.handle);
    // }
    //
    // public setPosition(x: number, y: number, z: number) {
    //   BlzSetSpecialEffectPosition(this.handle, x, y, z);
    // }
    //
    // public setRoll(roll: number) {
    //   BlzSetSpecialEffectRoll(this.handle, roll);
    // }
    //
    // public setScaleMatrix(x: number, y: number, z: number) {
    //   BlzSetSpecialEffectMatrixScale(this.handle, x, y, z);
    // }
    //
    // public setTime(value: number) {
    //   BlzSetSpecialEffectTime(this.handle, value);
    // }
    //
    // public setTimeScale(timeScale: number) {
    //   BlzSetSpecialEffectTimeScale(this.handle, timeScale);
    // }

    // public setYaw(y: number) {
    //   BlzSetSpecialEffectYaw(this.handle, y);
    // }

    /**
     * 新建特效(创建到坐标)
     */
    public static createEffect(modelName: string, x: number, y: number) {
        return new Effect(modelName, x, y);
    }

    /**
     * 新建特效(创建到坐标)
     */
    public static createEffectTarget(modelName: string, a: Widget, b: string) {
        return new Effect(modelName, a, b);
    }

    /**
     * 设置特效大小
     */
    public set effectSize(size: number) {
        EXSetEffectSize(this.handle, size)
        SetPariticle2Size(this.handle, size)
        this._data.size = size
    }

    /**
     * 获取特效大小
     */
    public get effectSize() {
        let size = this._data.size
        if (!size) {
            size = EXGetEffectSize(this.handle)
        }
        return size
    }

    /**
     * 设置特效高度
     */
    public set effectZ(Z: number) {
        EXSetEffectZ(this.handle, Z)
        this._data.z = Z
    }

    /**
     * 获取特效高度
     */
    public get effectZ() {
        let z = this._data.z
        if (!z) {
            z = EXGetEffectZ(this.handle)
        }
        return z
    }

    /**
     * 设置特效播放速度
     */
    public set effectSpeed(Speed: number) {
        EXSetEffectSpeed(this.handle, Speed)
        this._data.Speed = Speed
    }

    /**
     * 获取特效播放速度
     */
    public get effectSpeed() {
        return this._data.Speed
    }

    /**
     * 设置特效颜色
     */
    public set effectColor(Color: number) {
        EXSetEffectColor(this.handle, Color)
        this._data.color = Color
    }

    /**
     * 获取特效颜色
     */
    public get effectColor() {
        let Color = this._data.color
        if (!Color) {
            Color = EXGetEffectColor(this.handle)
        }
        return Color
    }

    /**
     * 设置迷雾可见特效
     */
    public set effectFogVisible(fogvisible: boolean) {
        EXSetEffectFogVisible(this.handle, fogvisible)
        this._data.fogvisible = fogvisible
    }

    /**
     * 获取特效迷雾可见情况
     */
    public get effectFogVisible() {
        return this._data.fogvisible
    }

    /**
     * 设置特效可见
     */
    public set effectVisible(visible: boolean) {
        EXSetEffectVisible(this.handle, visible)
        this._data.visible = visible
    }

    /**
     * 设置特效可见
     */
    public get effectVisible() {
        return this._data.visible
    }

    /**
     * 设置特效Z轴旋转
     */
    public set effectRotateZ(z: number) {
        this.reset()
        EXEffectMatRotateZ(this.handle, z)
        this._data.rotateZ = z
    }

    /**
     * 获取特效Z轴旋转
     */
    public get effectRotateZ() {
        return this._data.rotateZ
    }

    /**
     * 设置特效X轴旋转
     */
    public set effectRotateX(x: number) {
        EXEffectMatRotateX(this.handle, x)
        this._data.rotateX = x
    }

    /**
     * 获取特效X轴旋转
     */
    public get effectRotateX() {
        return this._data.rotateX
    }

    /**
     * 设置特效Y轴旋转
     */
    public set effectRotateY(y: number) {
        EXEffectMatRotateY(this.handle, y)
        this._data.rotateY = y
    }

    /**
     * 获取特效Y轴旋转
     */
    public get effectRotateY() {
        return this._data.rotateY
    }

    /**
     * 移动特效到XY
     */
    public setEffectXY(x: number, y: number) {
        EXSetEffectXY(this.handle, x, y)
        this._data.x = x
        this._data.y = y
    }

    /**
     * 获取特效XY轴
     */
    public getEffectXY(): { x: number, y: number } {
        let xy = {x: this._data.x, y: this._data.y}
        if (!xy.x) {
            xy.x = EXGetEffectX(this.handle)
        }
        if (!xy.y) {
            xy.y = EXGetEffectY(this.handle)
        }
        return xy
    }

    /**
     * 设置特效绑定handle
     */
    public setBindEffect(han: handle, node: string) {
        BindEffect(han, node, this.handle)
        this._data.binding = han
    }

    /**
     * 获取特效绑定的handle
     */
    public getBindEffect() {
        return this._data.binding
    }

    /**
     * 解除特效绑定
     */
    public unBindEffect() {
        UnBindEffect(this.handle)
        this._data.binding = null
    }

    /**
     * 设置特效播放动画
     */
    public setEffectAnimation(aon: number) {
        EXSetEffectAnimation(this.handle, aon)
    }

    /**
     * 特效重置（清空特效旋转与缩放）
     */
    public reset() {
        EXEffectMatReset(this.handle)
        this._data.size = 1
        this._data.rotateX = 0
        this._data.rotateY = 0
        this._data.rotateZ = 0
    }

    /**
     * 摧毁特效
     */
    public destroy() {
        delete this._data
        DestroyEffect(this.handle);
    }

    public static fromHandle(handle: effect): Effect {
        return this.getObject(handle);
    }
}
